onnxruntime/core/framework/bfc_arena.h

```diff
@@ -85,19 +88,17 @@ class BFCArena : public IAllocator {
 
   void* Reserve(size_t size) override;
 
-  FencePtr CreateFence(const SessionState* session_state) override {
-    // arena always rely on its device allocator to create fence
-    return device_allocator_->CreateFence(session_state);
-  }
-
   void GetStats(AllocatorStats* stats) override;
 
   size_t RequestedSize(const void* ptr);
 
   size_t AllocatedSize(const void* ptr);
 
- private:
-  void* AllocateRawInternal(size_t num_bytes, bool dump_log_on_failure);
+  virtual StreamAwareArena* AsStreamAwareAreana() {
```

Comment:
Instead of having the base class know about and returning a derived type, something like this could be considered. It does have a reinterpret_cast but that is validated. 

```c++
enum ArenaType {
  BFCArena,
  StreamAwareArena
};

class BFCArena 
{
    virtual ArenaType GetArenaType() const { get { return arena_type_; } }
}

class StreamAwareArena : BFCArena{
    static StreamAwareArena* FromBFCArena(BFCArena& arena)
    {
        return arena.GetArenaType() == ArenaType::StreamAwareArena ? 
            reinterpret_cast<StreamAwareArena*>(&arena) : nullptr;
    }
}
```

Reply:
I have applied this pattern, thanks for the suggestion.